org.gecko.emf.exporter/src/org/gecko/emf/exporter/EMFExportOptions.java

```diff
@@ -21,23 +21,16 @@ public interface EMFExportOptions {
 	// locale to use
 	String OPTION_LOCALE = "LOCALE";
 
-	// export non-containment references, in addition to containment references (exported by default)
+	// export non-containment references, in addition to containment references
+	// (exported by default)
 	String OPTION_EXPORT_NONCONTAINMENT = "EXPORT_NONCONTAINMENT";
 
 	// extract and export metadata as additional sheets
 	String OPTION_EXPORT_METADATA = "EXPORT_METADATA";
 
-	// automatically adjust column width based on contents
-	String OPTION_ADJUST_COLUMN_WIDTH = "ADJUST_COLUMN_WIDTH";
-
-	// generate links for references
-	String OPTION_GENERATE_LINKS = "GENERATE_LINKS";
-	
 	// generate mapping table
 	String OPTION_ADD_MAPPING_TABLE = "ADD_MAPPING_TABLE";
-	
-	// TODO: freezing rows is currently not supported in SODS
-	// freeze header row
-//	String OPTION_FREEZE_HEADER_ROW = "FREEZE_HEADER_ROW";
 
+	// show URIs instead of IDs
+	String OPTION_SHOW_URIS = "SHOW_URIS";
```

Comment:
I was wrong on my view on what the URIs would look like, as I had been blinded by the way EMFJson constructs its refrences (see `org.eclipse.emfcloud.jackson.databind.ser.EcoreReferenceSerializer.getHRef(...)`). 

We can use it to control if any ref fields should be written, that point to a non containment reference outside our exported list.

Example:
```
SHOW_URIS == true
//result would be

family.id, name,  family.father_ref, family.mother_ref
```

```
SHOW_URIS == false
//result would be

family.id, name
```

Reply:
I added a separate option and called it `SHOW_URIS` because URIs are only displayed in `ref_` fields (if that option is turned on).
 
Underneath, IDs are still being used, because I would not be able to automate this properly / non-deterministic behavior (different formats, lack of URI / incomplete URI / lack of unique identifier / etc) would make it very difficult to work. That is why I did not agree with you and continued using IDs underneath so all other features can work properly. It is only at the last stage where data collected in matrices is output to specific format (CSV, XLSX, ODS, etc.) that URIs can be displayed instead of IDs and only in `ref_` fields, otherwise IDs are displayed.
 
Having hopefully clarified this enough, do you still wish to have this removed ? If you wish to have option to control whether `ref_` fields are shown, shouldn't we name this in a different way - e.g. `SHOW_REFS`? Logic for this would be much different than that for `SHOW_URIS` / there would be no reuse as it's at a different stage that this would need to be handled - i.e. both at the time of creating headers, then data as certain columns (`ref_`) would need to be skipped.

Reply:
Given what I said in my comment [here](https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/23#discussion_r1370485601): if the `OPTION_EXPORT_NONCONTAINMENT` is not true, we must only export Objects in our given List. This means that every Object outside of our resulting  exported document would become a ref field, as EMF usually does for non Containment References in that exact case. 

This Option should control if such References should be exported or not. `SHOW_REFS` as a replacement name for that would be a good idea, as we call them refs.

Reply:
Yeah, the whole containment/noncontainment thing in EMF is a bit foreign to the normal Java model. Lets try to explain and answer your questions. This also takes your [comment ](https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/23#discussion_r1370568408) into account.

First in general: For EMF a valid Object at the time of serialization must be contained somewhere. And where ever this place is, the root of such a containment hierarchy must be a Resource. So, in order to understand Noncontainment we would need to understand Containment.

Lets assume, that the Father Reference has containment set to true. and we have the following pseudo code:

```
Resource toSave = new Resource("file:/foo.bar");

Family simpsons = new Simposons();
Family adams = new Adams();

Person homer = new Homer();

simspons.setFather(homer);
adams.setFather(homer);

toSave.getContents().add(simspons);
toSave.getContents().add(adams);
toSave.getContents().add(homer);

toSave,save();

```

This would result in the following pseudo json

```
[
  { 
    type: "Family"
    name : "Simpsons"
  },
 { 
    type: "Family"
    name : "Adams"
  },
  { 
    type: "Person"
    name : "Homer"
  }
]
```

Thus, there can be only one "owner" of Homer. The moment you add homer to a Containment Reference, EMF will remove it from where ever he was owned before. The Resources Content List is a so called `ContainmentEList`, which means anything you add there will be removed on any other containment.

Now we assume again, that Father is non Containment and repeat the whole thing:

```
Resource toSave = new Resource("file:/foo.bar");

Family simpsons = new Simposons();
Family adams = new Adams();

Person homer = new Homer();

simspons.setFather(homer);
adams.setFather(homer);

toSave.getContents().add(simspons);
toSave.getContents().add(adams);
toSave.getContents().add(homer);

toSave,save();

```

This would result in the following pseudo json

```
[
  { 
    type: "Family"
    name : "Simpsons"
   father.ref : "Homer"
  },
 { 
    type: "Family"
    name : "Adams"
    father.ref : "Homer"
  },
  { 
    type: "Person"
    name : "Homer"
  }
]
```

Now both Families point to Homer in the same Document.

Other Version:

```
Resource toSave = new Resource("file:/foo.bar");

Family simpsons = new Simposons();
Family adams = new Adams();

Resource homerResource = new Resource("file:/homer.bar");

Person homer = new Homer();

simspons.setFather(homer);
adams.setFather(homer);

toSave.getContents().add(simspons);
toSave.getContents().add(adams);
homerResource .getContents().add(homer);

toSave,save();

```

This would result in the following pseudo json

```
[
  { 
    type: "Family"
    name : "Simpsons"
   father.ref : "file:/homer.bar#Homer"
  },
 { 
    type: "Family"
    name : "Adams"
    father.ref : "file:/homer.bar#Homer"
  },
]
```

Now it points to the Homer Resource outside of the Document (may not be saved yet).



# Now to your specific Questions

> So, if I pass list of org.gecko.emf.osgi.example.model.basic.Person to org.gecko.emf.exporter.EMFExporter.exportEObjectsTo(List<EObject>, OutputStream, Map<?, ?>) and specify EXPORT_NONCONTAINMENT as false, wouldn't that still control only if non-contained (Address, Person) references are exported ? Because that is exactly as per EMF's definition of such.

That is correct. Only the Person with its contained Contacts and and Tags would be exported.

> Would SHOW_REFS in turn control in such case whether any references (contained or non-contained) are extracted and exported ? So, if I pass list of org.gecko.emf.osgi.example.model.basic.Person to org.gecko.emf.exporter.EMFExporter.exportEObjectsTo(List<EObject>, OutputStream, Map<?, ?>) and specify EXPORT_NONCONTAINMENT as false and SHOW_REFS as false, the resulting export would export only EAttribute structural features defined on org.gecko.emf.osgi.example.model.basic.Person ? (no EReference structural features).

No. Take my example from above as a baseline and add a mother as a containment feature:

```
Resource toSave = new Resource("file:/foo.bar");

Family simpsons = new Simposons();
Family adams = new Adams();

Resource homerResource = new Resource("file:/homer.bar");

Person homer = new Homer();

simspons.setFather(homer);
adams.setFather(homer);

Person marge = new Marge();
simpsons.setMother(marge);

toSave.getContents().add(simspons);
toSave.getContents().add(adams);
homerResource .getContents().add(homer);

toSave,save("SHOW_REFS ",  false);

```

This would result in the following pseudo json

```
[
  { 
    type: "Family"
    name : "Simpsons"
    mother: { 
      type : "Person",  
      name: "marge"
   }
 },
 { 
    type: "Family"
    name : "Adams"
  },
]
```

and also: 

```
Resource toSave = new Resource("file:/foo.bar");

Family simpsons = new Simposons();
Family adams = new Adams();

Person homer = new Homer();

Resource gomezResource = new Resource("file:/gomez.bar");

Person gomez= new Gomez();

simspons.setFather(homer);
adams.setFather(gomez);

Person marge = new Marge();
simpsons.setMother(marge);

toSave.getContents().add(simspons);
toSave.getContents().add(adams);
toSave.getContents().add(homer);

toSave,save("SHOW_REFS ",  false);

```

This would result in the following pseudo json

```
[
  { 
    type: "Family"
    name : "Simpsons"
    father.ref : "Homer"
    mother: { 
      type : "Person",  
      name: "marge"
   }
 },
 { 
    type: "Family"
    name : "Adams"
  },
  { 
    type: "Person"
    name : "Homer"
  }
]
```

Thus "SHOW_REFS" false mean: Don't show Reference URIs to Objects that are not contained in any of the given Object or the Resource itself (see the gomze, that is not Contained in the same Resource.

Reply:
@juergen-albert

I realize you've worked with EMF for much longer, but EMF is quite simple, at least in the sense of how containment / non-containment is defined on the EMF model itself (`<eStructuralFeatures xsi:type="ecore:EReference" (...) containment="true" (...)>`), which can then be checked via `org.eclipse.emf.ecore.EReference.isContainment()`. 

You referred several times to `Resource` and, as I mentioned in previous comment (https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/23#discussion_r1370568408), "_Perhaps Resource has a different understanding of what containement means, and you're referring to that implicitly_".

There are two public methods defined on `org.gecko.emf.exporter.EMFExporter` (as reminded in https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/23#discussion_r1370569897): 

 * `org.gecko.emf.exporter.EMFExporter.exportEObjectsTo(List<EObject>, OutputStream, Map<?, ?>)` and 
 
 * `org.gecko.emf.exporter.EMFExporter.exportResourceTo(Resource, OutputStream, Map<?, ?>)`
 
When I pass a list of `EObject`s to `org.gecko.emf.exporter.EMFExporter.exportEObjectsTo(List<EObject>, OutputStream, Map<?, ?>)`, where from would I be able to determine all these additional criterions which you mention? There, I only have access to `org.eclipse.emf.ecore.EReference.isContainment()` (derived from whatever is defined in the EMF model itself) and `exportOptions` (in this particular case, options `EXPORT_NONCONTAINMENT` and `SHOW_REFS`).

As mentioned in https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/23#discussion_r1370569897, perhaps you wish to change implementation of `org.gecko.emf.exporter.EMFExporter.exportResourceTo(Resource, OutputStream, Map<?, ?>)` method only, i.e. such cases should only be handled if dealing with `Resource` (`org.gecko.emf.exporter.EMFExporter.exportResourceTo(Resource, OutputStream, Map<?, ?>)` method), not with `EObject` list (`org.gecko.emf.exporter.EMFExporter.exportEObjectsTo(List<EObject>, OutputStream, Map<?, ?>)` method) ? If I can programmatically access / derive these criterions from `Resource` passed to `org.gecko.emf.exporter.EMFExporter.exportResourceTo(Resource, OutputStream, Map<?, ?>)`, to properly automate handling such cases, that is.

Reply:
> Perhaps Resource has a different understanding of what containement means, and you're referring to that implicitly

Yes, that is kind of correct. EObjects must be contained somewhere. Either directly in a resource or in an EObject with a proper containment reference, where the root is contained in a Resource. The basic check is: `EObject.eResource()` must return something. 

Looking at the the two methods you have I can say the following:

`org.gecko.emf.exporter.EMFExporter.exportResourceTo(Resource, OutputStream, Map<?, ?>)` - here it is fairly easy as you would only need to check if any Object you may encounter points to the given resource.

`org.gecko.emf.exporter.EMFExporter.exportEObjectsTo(List<EObject>, OutputStream, Map<?, ?>)` - here it is a bit harder. If you encounter a non containment Reference, you would need to get the Object it points to, without triggering a proxy resolve and then you would need to check if the EObject you have received is in the initial List you have gotten to export. If it is a Proxy, you would need to compare its URI, with the URIs of all the given Objects in your List.

The `exportResourceTo`, is the one I and the Customer really wants and needs. 
I see the other one as useful for testing, but I have the feeling, that it otherwise creates more complexity.